tests/driver_apds99xx_full/main.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2018 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @brief       Test application for Vishay APDS99XX proximity and ambient light sensor
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ * @file
+ *
+ * The test application demonstrates the use of the APDS99XX driver with
+ * interrupts.
+ *
+ * - Ambient light sensing (ALS) interrupt is used as data-ready interrupt
+ *   which generates an interrupt in each cycle when the ALS has been finished.
+ *
+ * - Proximity sensing (PRX) interrupt is generated when the count value
+ *   exceeds the threshold of 200. It doesn't use the low threshold.
+ *
+ * The application uses the following configurations:
+ *
+ * - 1 x ALS gain,
+ * - 64 steps as ALS integration time to use the full range of uint16_t,
+ * - 8 IR LED pulses at 100 mA current for proximity sensing,
+ * - 1 x proximity sensing gain, and,
+ * - no waits.
+ *
+ * Depending on the sensor, a measurement cycle takes from 156 ms (APDS9950)
+ * to 179 ms (APDS9960).
+ */
+
+#include <stdio.h>
+
+#include "thread.h"
+#include "thread_flags.h"
+#include "xtimer.h"
+
+#include "apds99xx.h"
+#include "apds99xx_params.h"
+
+/* thread flags are used to indicate interrupt events to a waiting thread */
+#define APDS99XX_IRQ_FLAG  0x1000
+
+/* reference to main thread */
+thread_t* t_main;
+
+static void apds99xx_isr (void *arg)
+{
+    /*
+     * This ISR function is executed in the interrupt context. It must not be
+     * blocking or time-consuming and must not access the sensor directly
+     * via I2C.
+     *
+     * Therefore, the ISR function only indicates to the waiting thread that
+     * an interrupt has occurred which needs to be handled in the thread
+     * context.
+     *
+     * For this purpose, a message could be sent to the waiting thread.
+     * However, sending a message in an ISR requires a message queue in
+     * the waiting thread. Since it is not relevant how many interrupts
+     * have occurred since last interrupt handling, but only that an
+     * interrupt has occurred, we simply use a thread flag here instead of
+     * a message for simplicity.
+     */
+    (void)arg;
+    thread_flags_set(t_main, APDS99XX_IRQ_FLAG);
+}
+
+int main(void)
+{
+    apds99xx_t dev;
+
+    /* save the reference to the main thread */
+    t_main = (thread_t*)sched_threads[sched_active_pid];
```

Comment:
better use `t_main = thread_getpid();` here?

Reply:
The reference to the main thread is required in `apds99xx_isr`. Using `thread_get(pid)` doesn't work since it replaces the const attribute.

I changed the interrupt handling to a mutex based approach.